TraceInfrastructure/Passes/MemProfile.cpp

```diff
@@ -0,0 +1,160 @@
+                    lastBuilder.CreateCall(MemProfDestroy);
+                }
+                else if (auto resumeInst = dyn_cast<ResumeInst>(BB->getTerminator()))
+                {
+                    auto endInsertion = BB->getTerminator();
+                    auto *lastInst = cast<Instruction>(endInsertion);
+                    IRBuilder<> lastBuilder(lastInst);
+                    lastBuilder.CreateCall(MemProfDestroy);
+                }
+                else if (auto unreachableInst = dyn_cast<UnreachableInst>(BB->getTerminator()))
+                {
+                    auto endInsertion = BB->getTerminator();
+                    auto *lastInst = cast<Instruction>(endInsertion);
+                    IRBuilder<> lastBuilder(lastInst);
+                    lastBuilder.CreateCall(MemProfDestroy);
+                }
+            }
+            for (auto bi = BB->begin(); bi != BB->end(); bi++)
+            {
+                if (auto call = dyn_cast<CallBase>(bi))
+                {
+                    if (call->getCalledFunction())
+                    {
+                        if (call->getCalledFunction()->getName() == "exit")
+                        {
+                            IRBuilder<> destroyInserter(call);
+                            auto insert = destroyInserter.CreateCall(MemProfDestroy);
+                            insert->moveBefore(call);
+                            call->setDebugLoc(NULL);
+                        }
+                    }
+                }
+            }
+
+            auto *block = cast<BasicBlock>(BB);
+            auto dl = block->getModule()->getDataLayout();
+            int64_t blockId = GetBlockID(block);
+
+            for (BasicBlock::iterator BI = block->begin(), BE = block->end(); BI != BE; ++BI)
+            {
+
+                auto *CI = dyn_cast<Instruction>(BI);
+                std::vector<Value *> values;
+                if (auto *load = dyn_cast<LoadInst>(CI))
+                {
+                    IRBuilder<> builder(load);
+                    Value *addr = load->getPointerOperand();
+                    auto *type = load->getType();
+                    uint64_t dataSize = dl.getTypeAllocSize(type);
+
+                    // addr
+                    auto castCode = CastInst::getCastOpcode(addr, true, PointerType::get(Type::getInt8PtrTy(block->getContext()), 0), true);
+                    Value *addrCast = builder.CreateCast(castCode, addr, Type::getInt8PtrTy(block->getContext()));
+                    values.push_back(addrCast);
+                    //bb id
+                    Value *blockID = ConstantInt::get(Type::getInt64Ty(BB->getContext()), (uint64_t)blockId);
+                    values.push_back(blockID);
+                    // data size
+                    Value *dataSizeValue = ConstantInt::get(Type::getInt64Ty(BB->getContext()), dataSize);
+                    values.push_back(dataSizeValue);
+                    // type 0 = load, 1 = store
```

Comment:
Here you could consider making two separate functions. One for load one for store. That way you will compartmentalize your code more effectively